### PR TITLE
[Refact] (serde) refact mysql serde with data type

### PR DESCRIPTION
**Pull request by amorynan**
# Proposed changes
refact mysql output (de)serialize with data type serde , avoid accoriding switch case Primitive type writed in mysqlWriter  

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/data_types/serde/data_type_decimal_serde.h [hunk 1]
```diff
@@ -60,8 +61,49 @@ class DataTypeDecimalSerDe : public DataTypeSerDe {
...
+                                  int row_idx, int start, int end, int scale, bool col_const) const;
 };
 
+template <typename T>
+template <bool is_binary_format>
+Status DataTypeDecimalSerDe<T>::_write_column_to_mysql(
+        const IColumn& column, std::vector<MysqlRowBuffer<is_binary_format>>& result, int row_idx,
+        int start, int end, int scale, bool col_const) const {
+    int buf_ret = 0;
+    auto& data = static_cast<const ColumnDecimal<T>&>(column).get_data();
```
Comment (zclllyybb) [inline be/src/vec/data_types/serde/data_type_decimal_serde.h:89]: please make sure all uses like this are of form `assert_cast`

File: be/src/vec/data_types/serde/data_type_array_serde.cpp [hunk 2]
```diff
@@ -83,5 +84,36 @@ void DataTypeArraySerDe::read_column_from_arrow(IColumn& column, const arrow::Ar
...
+    int buf_ret = 0;
+    auto& column_array = assert_cast<const ColumnArray&>(column);
+    auto& offsets = column_array.get_offsets();
+    auto& data = column_array.get_data();
+    for (ssize_t i = start; i < end; ++i) {
+        if (0 != buf_ret) {
+            return Status::InternalError("pack mysql buffer failed.");
+        }
+        const auto col_index = index_check_const(i, col_const);
+        LOG(WARNING) << "amory: " << col_index;
```
Comment (zclllyybb) [inline be/src/vec/data_types/serde/data_type_array_serde.cpp:100]: remember to remove it~

**Review by zclllyybb — COMMENTED** [review]
looks good overall, need a little update.

**Comment by amorynan** [comment]
run buildall

**Comment by yiguolei** [comment]
run buildall

**Review by zclllyybb — DISMISSED** [review]
LGTM

**Review by eldenmoon — DISMISSED** [review]
LGTM

**Comment by amorynan** [comment]
run buildall

File: be/src/vec/data_types/serde/data_type_nullable_serde.cpp [hunk 3]
```diff
@@ -131,5 +132,30 @@ void DataTypeNullableSerDe::read_column_from_arrow(IColumn& column, const arrow:
...
+    int buf_ret = 0;
+    auto& col = static_cast<const ColumnNullable&>(column);
+    auto& nested_col = col.get_nested_column();
+    for (ssize_t i = start; i < end; ++i) {
+        LOG(WARNING) << "amory: " << start << " : " << end;
+        if (0 != buf_ret) {
+            return Status::InternalError("pack mysql buffer failed.");
+        }
+        const auto col_index = index_check_const(i, col_const);
+        LOG(WARNING) << "amory: " << col_index << " is null: " << col.is_null_at(col_index);
```
Comment (eldenmoon) [inline be/src/vec/data_types/serde/data_type_nullable_serde.cpp:148]: same as above

File: be/src/vec/data_types/serde/data_type_nullable_serde.cpp [hunk 4]
```diff
@@ -131,5 +132,30 @@ void DataTypeNullableSerDe::read_column_from_arrow(IColumn& column, const arrow:
...
 
+template <bool is_binary_format>
+Status DataTypeNullableSerDe::_write_column_to_mysql(
+        const IColumn& column, std::vector<MysqlRowBuffer<is_binary_format>>& result, int row_idx,
+        int start, int end, int scale, bool col_const) const {
+    int buf_ret = 0;
+    auto& col = static_cast<const ColumnNullable&>(column);
+    auto& nested_col = col.get_nested_column();
+    for (ssize_t i = start; i < end; ++i) {
+        LOG(WARNING) << "amory: " << start << " : " << end;
```
Comment (eldenmoon) [inline be/src/vec/data_types/serde/data_type_nullable_serde.cpp:143]: remove this log

File: be/src/vec/data_types/serde/data_type_nullable_serde.cpp [hunk 5]
```diff
@@ -131,5 +132,30 @@ void DataTypeNullableSerDe::read_column_from_arrow(IColumn& column, const arrow:
...
+        if (0 != buf_ret) {
+            return Status::InternalError("pack mysql buffer failed.");
+        }
+        const auto col_index = index_check_const(i, col_const);
+        LOG(WARNING) << "amory: " << col_index << " is null: " << col.is_null_at(col_index);
+        if (col.is_null_at(col_index)) {
```
Comment (eldenmoon) [inline be/src/vec/data_types/serde/data_type_nullable_serde.cpp:149]: why not `col.is_null_at(row_idx)`?

File: be/src/vec/sink/vmysql_result_writer.cpp [hunk 6]
```diff
@@ -624,269 +625,8 @@ Status VMysqlResultWriter<is_binary_format>::append_block(Block& input_block) {
...
-            }
-            break;
-        }
-        default: {
-            LOG(WARNING) << "can't convert this type to mysql type. type = "
-                         << _output_vexpr_ctxs[i]->root()->type();
-            return Status::InternalError("vec block pack mysql buffer failed.");
-        }
-        }
+        RETURN_IF_ERROR(type_ptr->get_serde()->write_column_to_mysql(*column_ptr, rows_buffer, 0, 0,
```
Comment (eldenmoon) [inline be/src/vec/sink/vmysql_result_writer.cpp:628]: get_serde will return temporary DataTypeSerDeSPtr and may lead to bad performace

Comment (amorynan) [reply be/src/vec/sink/vmysql_result_writer.cpp:628]: yes , later I would use create_data_type_serdes

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.55 seconds
 stream load tsv:          418 seconds loaded 74807831229 Bytes, about 170 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          58 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          79.2 seconds inserted 10000000 Rows, about 126K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230525042436_clickbench_pr_150119.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.02 seconds
 stream load tsv:          420 seconds loaded 74807831229 Bytes, about 169 MB/s
 stream load json:         23 seconds loaded 2358488459 Bytes, about 97 MB/s
 stream load orc:          58 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          77.1 seconds inserted 10000000 Rows, about 129K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230525044731_clickbench_pr_150128.html

**Comment by amorynan** [comment]
run buildall

File: regression-test/data/dynamic_table_p0/sql/q02.out [hunk 7]
```diff
@@ -18,15 +18,15 @@
 4060
 
 -- !q02_7 --
-['.805', '.621']
```
Comment (eldenmoon) [inline regression-test/data/dynamic_table_p0/sql/q02.out:21]: the behavior seems different with previous, what if user use the previous behavior? will this lead to incompatible？

Comment (amorynan) [reply regression-test/data/dynamic_table_p0/sql/q02.out:21]: this is only in array string , for standard in json , ['aa', 'bb'] is wrong expr, and I met a user issue for changing the array output, I think we should make array string is right!

**Comment by amorynan** [comment]
run p0

**Review by eldenmoon — DISMISSED** [review]
LGTM

**Comment by amorynan** [comment]
run p0

**Comment by amorynan** [comment]
run p0

**Comment by amorynan** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.56 seconds
 stream load tsv:          421 seconds loaded 74807831229 Bytes, about 169 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          79.8 seconds inserted 10000000 Rows, about 125K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230526030635_clickbench_pr_150923.html

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**